tools/mockgen/mockgen.go

```diff
@@ -0,0 +1,106 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package mockgen contains a wrapper for the `mockgen` tool, which bypasses running `mockgen` if the destination file
+// is newer than the source file. This is useful because `mockgen` is slow and can be a bottleneck in the build process.
+package mockgen
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"syscall"
+)
+
+// Run a cached version of `mockgen`, which checks the modification times of the source and destination files,
+// running mockgen only if necessary. This is similar to the behavior of `Make`, but it wasn't easy to express this
+// in a Makefile because the generate commands are in our source code.
```

Comment:
If we can guarantee that all generated files are named something like `$FILENAME_mock.go` you can use make for this; that's exactly what I did at a previous job

You can do the same if they're generated to `mocks/$FILENAME.go`

Reply:
That's what I wanted to do originally, but what would we do with go:generate statements that aren't for mockgen? I think that the only thing we'd be leveraging from Make is the time comparison, but that logic is so simple that I don't think we lose much by duplicating it.

Reply:
grep for `go:generate\s+mockgen` and we wouldn't need to care.

This works though, but I wish we'd use make properly if we're going to use it at all